### PR TITLE
Pasteboard,Keybindhint & some minor fixes

### PR DESCRIPTION
**Pull request by kRHYME7**
# Pull Request

## Description

Vertical Monitor Support for cliphist.sh
+ Respects scale and transform values
+ Respects waybar pos and use it as boundary 
+ follows width from the rasi file but scaling should be a ratio of width [ this case is (20/55) or ~ (36/100)  ] 
+ Majority of the operations are done inside the json data(easier)
+ 

## Type of change

Please put an `x` in the boxes that apply:

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that would cause existing functionality to not work as expected)
- [ ] **Documentation update** (non-breaking change; modified files are limited to the documentations)
- [x] **Technical debt** (a code change that does not fix a bug or add a feature but makes something clearer for devs)
- [ ] **Other** (provide details below)

## Screenshots

240417_00h43m31s_screenshot

240417_00h44m08s_screenshot

(if appropriate)

### CODE REVIEW

File: Configs/.local/share/bin/cliphist.sh [hunk 1]
```diff
@@ -1,86 +1,90 @@
...
-clip_h=$(( y_mon*clip_h/100 ))
-max_x=$((x_mon - clip_w - 5 )) #offset of 5 for gaps
-max_y=$((y_mon - clip_h - 15 )) #offset of 15 for gaps
-x_cur=$((x_cur - x_offset))
-y_cur=$((y_cur - y_offset))
-# 
-x_cur=$(( x_cur < min_x ? min_x : ( x_cur > max_x ? max_x :  x_cur)))
-y_cur=$(( y_cur < min_y ? min_y : ( y_cur > max_y ? max_y :  y_cur)))
+#? Parse clipboard.rasi and fetch the width. Should consider percent
+clpWidth=$(awk '/window {/,/}/ { if (/width:/) { split($2, a, "%"); print a[1] } }' "${roconf}" )
```
Comment (prasanthrangan) [inline Configs/.local/share/bin/cliphist.sh:12]: I still don't remember what i was on, to set window width in %  :)
I think it would be better to use `em` unit for width and height.

Comment (kRHYME7) [reply Configs/.local/share/bin/cliphist.sh:12]: Okie, As I have understood this, if we set it to em it will scale to the set font size right?

Comment (prasanthrangan) [reply Configs/.local/share/bin/cliphist.sh:12]: > Okie, As I have understood this, if we set it to em it will scale to the set font size right?

yup, setting it to em will scale the menu based on font size.. so for example, if we set font size as `10` and width as `30em` then it will scale to `300px` on screen... so we dont care about the monitor specs...

Comment (kRHYME7) [reply Configs/.local/share/bin/cliphist.sh:12]: Well, try to adjust on how this treats the boundaries. But Goodnight for now 😂

Comment (prasanthrangan) [reply Configs/.local/share/bin/cliphist.sh:12]: > Well, try to adjust on how this treats the boundaries. But Goodnight for now 😂

once you have fixed size of rofi window... it should easy to do that, like windows does... take the cursor position and check the below and above... and offset the rofi window accordingly

Comment (kRHYME7) [reply Configs/.local/share/bin/cliphist.sh:12]: So what happens now is I got unknown issues. Does rofi adheres to the em unit * fontsize ?

image


So I'm using the default fnt size : 9 and 20em for W and 30 em for H
So hyprctl reports that the rofi layer if we try to see the ratio uses 8.5 as fnt size.  which ruins the boundaries. ( lazy to debug which part the .rasi does this but I am certain the scripts follows the em * fnt  principle )


For now, to solve this I still end up overwriting the width and height.

Will push this later for review and revision

Comment (kRHYME7) [reply Configs/.local/share/bin/cliphist.sh]: I also tried to get the equivalents in percent.

File: Configs/.local/share/bin/cliphist.sh [hunk 2]
```diff
@@ -1,86 +1,87 @@
...
+export curXpos=\(.x - $monXpos - $x_offset )
+export curYpos=\(.y - $monYpos - $y_offset)
+"')"
 
+# TODO Dynamically follow the rofi scaling if .rasi changes
+# Limit board location to monitor
+wBoard=${wBoard:-300}
+hBoard=${hBoard:-550}
+xBound=$((monWidth - wBoard - wBar))
+yBound=$((monHeight - hBoard - hBar))
```
Comment (kRHYME7) [inline Configs/.local/share/bin/cliphist.sh:48]: This part is our magic number. For now, this is not dynamic.

Comment (prasanthrangan) [reply Configs/.local/share/bin/cliphist.sh:48]: > This part is our magic number. For now, this is not dynamic.

whats the `wBoard` and `hBoard` value for??

Comment (kRHYME7) [reply Configs/.local/share/bin/cliphist.sh:48]: The width and height of the pasteboard

image

It's blank for now, but ready to be overwritten

Comment (kRHYME7) [reply Configs/.local/share/bin/cliphist.sh]: Added a logic to just save the scaling
image

File: Configs/.local/share/bin/swwwallbash.sh
Comment (kRHYME7) [inline Configs/.local/share/bin/swwwallbash.sh:1]: Added so we can use the wallbash engine and get the image path for user specific things

File: Configs/.local/share/bin/keybinds_hint.sh
Comment (kRHYME7) [inline Configs/.local/share/bin/keybinds_hint.sh:1]: Added overrides  for height width and the delimeter to use

File: Configs/.config/rofi/clipboard.rasi [hunk 3]
```diff
@@ -10,8 +10,8 @@ configuration {
 
 // Main //
 window {
-    height:                      55%;
-    width:                       20%;
+    width:                       16.7em;
+    height:                      30.6em;
```
Comment (kRHYME7) [inline Configs/.config/rofi/clipboard.rasi:14]: Equivalent to above

**Review by rubiin — APPROVED** [review]
LGTM

**Review by prasanthrangan — APPROVED** [review]
this seems like we are doing way too much... I think this could be just as simple as getting the cursor position then anchoring/offsetting the rofi window relative to cursor position.

Let me merge this.. and I'll try to simplify this as much as I can!

File: Configs/.local/share/bin/swwwallbash.sh [hunk 4]
```diff
@@ -5,7 +5,7 @@
 
 export scrDir="$(dirname "$(realpath "$0")")"
 source "${scrDir}/globalcontrol.sh"
-wallbashImg="${1}"
+export wallbashImg="${1}"
```
Comment (prasanthrangan) [inline Configs/.local/share/bin/swwwallbash.sh:8]: wait why is this required? for user specific things, the `${theme}/wall.set` should give you that.

Comment (kRHYME7) [reply Configs/.local/share/bin/swwwallbash.sh:8]: Is this file available during the wallbash loop? use case is for example auto converting the wallpaper to something.

Comment (prasanthrangan) [reply Configs/.local/share/bin/swwwallbash.sh:8]: hey @kRHYME7 , i kinda updated the logic for clipboard position... works fine is my system... can you please give it a try for multi/vert monitors?

Comment (kRHYME7) [reply Configs/.local/share/bin/swwwallbash.sh:8]: sure no probs

Comment (kRHYME7) [reply Configs/.local/share/bin/swwwallbash.sh:8]: Almost there, But the boundaries did not respect if my scaling is not equal to 1. If it's 1 then all good

Comment (rubiin) [reply Configs/.local/share/bin/swwwallbash.sh:8]: I tried the clipboard no probs so far as well

Comment (kRHYME7) [reply Configs/.local/share/bin/swwwallbash.sh:8]: Ops wait lemme screencord this. Saw some multiple issues

Comment (rubiin) [reply Configs/.local/share/bin/swwwallbash.sh:8]: the positioning is working fine so as the scaling on my side

Comment (prasanthrangan) [reply Configs/.local/share/bin/swwwallbash.sh:8]: > Almost there, But the boundaries did not respect if my scaling is not equal to 1. If it's 1 then all good

oh noooo.... what was i thinking... completely forgot to factor in scaling :)
fixing it now!

Comment (kRHYME7) [reply Configs/.local/share/bin/swwwallbash.sh:8]: Not much and if this is working at normal circumstances then it should be fine IG. 

Becomes too wide for my 3000*2000 monitor (scaled to 2)
image

image

image

Monitors
image

image


image

Does not handle if monitors are mapped oddly dynamically
image

image

Comment (prasanthrangan) [reply Configs/.local/share/bin/swwwallbash.sh:8]: > Not much and if this is working at normal circumstances then it should be fine IG.
> 
> Becomes too wide for my 3000*2000 monitor (scaled to 2)
> 

Fixed now, please try the latest commit... it should factor in scaling now...
regarding width, i did increase it to increase the field size.

Comment (kRHYME7) [reply Configs/.local/share/bin/swwwallbash.sh:8]: Still got this 
image

Also this 

image

But maybe this is an odd setup?  HAHA
